plugins: container-rootfs-relative paths

tiborvass opened the pull request:
Legacy plugins expect host-relative paths (such as for Volume.Mount).
However, a containerized plugin cannot respond with a host-relative
path. Therefore, this commit modifies new volume plugins' paths in Mount
and List to prepend the container's rootfs path.

Volume plugins that do mounts and want it to propagate to the host namespace, need to mount inside `/mnt`.

Signed-off-by: Tibor Vass tibor@docker.com

tiborvass commented on plugin/manager.go:
```diff
@@ -58,7 +58,7 @@ func Init(root string, remote libcontainerd.Remote, rs registry.Service, liveRes
 	manager = &Manager{
 		libRoot:           root,
-		runRoot:           "/run/docker",
+		runRoot:           "/run/docker/plugins",
```
@anusha-ragunathan @vieux I changed this to accommodate legacy plugins that have hardcoded `/run/docker/plugins` already. I had initially chosen `/run/docker` because it was shorter.

anusha-ragunathan replied:
Works for me. It reduces the code change required by the legacy plugins to move to pluginv2.

anusha-ragunathan replied:
@tiborvass : you dont need to change this for plugins. This is the host directory that gets bind mounted to the plugin. The plugin run time path is `defaultPluginRuntimeDestination` which is already set to `/run/docker/plugins` https://github.com/docker/docker/blob/fa7c1a68a68ffb7c6851c59d2c0b41091b9d6e5e/plugin/v2/plugin.go#L29

We could change runRoot back to `/run/docker/plugins` to indicate that the directories that get created under there are plugins related.

cpuguy83 replied:
For a separate PR, but this should not be hard-coded to `/run/docker` but instead use the daemon `exec root`

anusha-ragunathan replied:
The reason to hard code is that, this holds the plugin socket. There's a socket path limitation on Linux to 108 bytes. The daemon exec root doesnt have any socket and is not bound by this limitation. So we have to choose a path that always works.

tiborvass commented on pkg/plugins/plugins.go:
```diff
@@ -73,6 +73,12 @@ type Plugin struct {
 
+// PathsRelativeTo returns the path to which all paths returned by the plugin are relative to.
+// For legacy plugins, this is always relative to the host's root directory.
+func (p *Plugin) PathsRelativeTo() string {
```
I realize this name sucks, so alternatives are welcome. Rootfs? Basepath? Hostpath?

vieux replied:
I like Basepath

anusha-ragunathan replied:
PluginRoot is accurate.

tiborvass replied:
@anusha-ragunathan the problem is that for legacy plugin the notion of "pluginroot" is kind of moot. This is returning the "basepath" to which all paths returned by the plugin will be relative to.

anusha-ragunathan replied:
Its not moot. It's still the legacy plugins' rootfs.

anusha-ragunathan commented on plugin/v2/plugin.go:
```diff
@@ -205,18 +213,53 @@ func (p *Plugin) GetTypes() []types.PluginInterfaceType {
 // InitSpec creates an OCI spec from the plugin's config.
 func (p *Plugin) InitSpec(s specs.Spec, libRoot string) (*specs.Spec, error) {
-	rootfs := filepath.Join(libRoot, p.PluginObj.ID, "rootfs")
+	p.Rootfs = filepath.Join(libRoot, p.PluginObj.ID, "rootfs")
```
p.Rootfs can also be used in https://github.com/docker/docker/blob/master/plugin/backend.go#L132

anusha-ragunathan commented:
Lets keep this PR specifically for container-rootfs-relative path fix and open new ones for the other bug fixes. Keeps a cleaner history.

anusha-ragunathan commented on plugin/v2/plugin.go:
```diff
@@ -205,18 +213,53 @@ func (p *Plugin) GetTypes() []types.PluginInterfaceType {
 		Readonly: false, // TODO: all plugins should be readonly? settable in manifest?
 	}
 
+	// TODO: the following should be only if granting device permissions
```
Lets fix this todo by adding to Manifest's privilege section. Something like "REQUEST_DEVICE_CREATION".

anusha-ragunathan commented on plugin/v2/plugin.go:
```diff
@@ -250,12 +302,14 @@ func (p *Plugin) InitSpec(s specs.Spec, libRoot string) (*specs.Spec, error) {
+
+	// TODO: what about duplicates?
+	// TODO: Should not need CAP_ prefix in manifest?
+	s.Process.Capabilities = append(s.Process.Capabilities, p.PluginObj.Manifest.Capabilities...)
```
Not your change, but we need some naming consistency here. "Capabilities" is overloaded:
1. plugin/store/store.go uses "Capability" to mean type of plugin. We should rename this to "pluginType" or simply "type".
2.  VolumeDriver.Capabilities is a endpoint for volume plugins. Nothing can be done here.
3. Manifest has a section for Linux capabilities (referenced here)

anusha-ragunathan commented on pkg/plugins/discovery.go:
```diff
@@ -33,6 +33,10 @@ func Scan() ([]string, error) {
 			return nil
 		}
 
+		if fi.IsDir() {
```
We allow nesting the socket inside subdirectories. 
From https://docs.docker.com/engine/extend/plugin_api/ : "You can define each plugin into a separated subdirectory if you want to isolate definitions from each other. For example, you can create the flocker socket under /run/docker/plugins/flocker/flocker.sock and only mount /run/docker/plugins/flocker inside the flocker container."

So dont do this.

anusha-ragunathan commented on plugin/store/store.go:
```diff
@@ -113,7 +113,7 @@ func (ps *PluginStore) getAllByCap(capability string) []CompatPlugin {
 	result := make([]CompatPlugin, 0, 1)
 	for _, p := range ps.plugins {
-		if _, err := p.FilterByCap(capability); err == nil {
+		if _, err := p.FilterByCap(capability); err == nil && p.PluginObj.Enabled {
```
Use p.IsEnabled(). Eventually, we can make PluginObj an internal object.

anusha-ragunathan replied:
We want to return only enabled plugins from LookupWithCapability. This applies to FindWithCapability as well. So lets leave this function alone and move the check for enabled in LookupWithCapability and FindWithCapability.

anusha-ragunathan commented on plugin/store/store.go:
```diff
@@ -180,7 +180,7 @@ func LookupWithCapability(name, capability string) (CompatPlugin, error) {
 		}
 		p, err = store.GetByName(fullName)
-		if err == nil {
+		if err == nil && p.PluginObj.Enabled {
```
p.IsEnabled()

anusha-ragunathan commented on plugin/v2/plugin.go:
```diff
@@ -241,6 +284,15 @@ func (p *Plugin) InitSpec(s specs.Spec, libRoot string) (*specs.Spec, error) {
 
+	defaultMounts := make([]specs.Mount, 0, len(s.Mounts))
+	for _, m := range s.Mounts {
+		if !strings.HasPrefix(m.Destination, "/dev/") && true { // TODO: && user specified /dev
```
How about delete the mount with "/dev/" prefix from s.Mounts rather than making copies?

for index, _ := range s.Mounts {
		if strings.HasPrefix(m.Destination...) {
			s.Mounts = append(s.Mounts[:index], s.Mounts[index+1:]...)
		}
	}

anusha-ragunathan commented on plugin/v2/plugin.go:
```diff
@@ -205,18 +213,53 @@ func (p *Plugin) GetTypes() []types.PluginInterfaceType {
 		Options:     []string{"rbind", "rshared"},
 	})
+
+	if p.PluginObj.Manifest.Network.Type != "" {
```
https://github.com/docker/docker/blob/3ea762b9f6ba256cf51bd2c35988f0c48bccf0b0/container/container_unix.go#L114 does the same thing for containers.

anusha-ragunathan replied:
Lets look to refactor. At the least, we should perform more error checking such as stating the host file before adding to mounts.

tiborvass commented on plugin/store/store.go:
```diff
@@ -113,7 +113,7 @@ func (ps *PluginStore) getAllByCap(capability string) []CompatPlugin {
 	result := make([]CompatPlugin, 0, 1)
 	for _, p := range ps.plugins {
-		if _, err := p.FilterByCap(capability); err == nil {
+		if _, err := p.FilterByCap(capability); err == nil && p.IsEnabled() {
```
@anusha-ragunathan I could not change this in the caller function because I have only access to CompatPlugin interface values, and not *Plugin objects.

anusha-ragunathan replied:
Fair enough.

anusha-ragunathan commented:
FAIL: docker_cli_daemon_experimental_test.go:177: DockerDaemonSuite.TestVolumePlugin

[d0f844bc9335c] waiting for daemon to start
[d0f844bc9335c] daemon started
docker_cli_daemon_experimental_test.go:223:
    c.Assert(err, checker.IsNil, check.Commentf(out))
... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc421c5b0e0), Stderr:[]uint8(nil)} ("exit status 127")
... Unable to find image 'busybox:latest' locally
latest: Pulling from library/busybox
8ddc19f16526: Pulling fs layer
8ddc19f16526: Download complete
8ddc19f16526: Pull complete
Digest: sha256:a59906e33509d14c036c8678d687bd4eec81ed7c4b8ce907b888c607f6a1e0e6
Status: Downloaded newer image for busybox:latest
/go/src/github.com/docker/docker/bundles/1.13.0-dev/binary-client/docker: Error response from daemon: oci runtime error: rootfs_linux.go:53: mounting "/tmp/data" to rootfs "/go/src/github.com/docker/docker/bundles/1.13.0-dev/test-integration-cli/d0f844bc9335c/root/vfs/dir/fe75f9766650efb2ecd73bee99b3c9eb67301379cccf3e119582c92016c3a318" caused "stat /go/src/github.com/docker/docker/bundles/1.13.0-dev/test-integration-cli/d0f844bc9335c/root/plugins/2666fee9e7017759705396aa5301f13af1385ca24e8665106e0cdcd687385816/rootfs/data/plugin-volume: no such file or directory".

ekristen commented:
Please let me know if I can assist. Looking forward to seeing this get merged for 1.13.

tiborvass commented:
Thanks, I'm planning to update this soon.

thaJeztah commented:
ping @tiborvass what's the status here?

crosbymichael commented:
@tiborvass ping

tiborvass commented:
@crosbymichael i'm trying to figure out a solution for the issue I'm hitting.

tiborvass commented:
need to fix leaking mounts

anusha-ragunathan commented:
Failing test in experimental. With a rebase, janky will fail. 
```
 FAIL: docker_cli_daemon_experimental_test.go:114: DockerDaemonSuite.TestDaemonShutdownLiveRestoreWithPlugins
23:47:19 
23:47:19 [d3a0ed9d8ce23] waiting for daemon to start
23:47:19 [d3a0ed9d8ce23] daemon started
23:47:19 docker_cli_daemon_experimental_test.go:141:
23:47:19     c.Fatalf("Expected exit code '0', got %d err: %v output: %s ", ec, err, out)
23:47:19 ... Error: Expected exit code '0', got 1 err: exit status 1 output:  
23:47:19 
23:47:19 [d3a0ed9d8ce23] waiting for daemon to start
23:47:19 [d3a0ed9d8ce23] daemon started
23:47:19 [d3a0ed9d8ce23] exiting daemon
```
Also, PR 
- needs a rebase.
- commit on `plugins: do not try to contact disabled plugin` can be removed, since it has been fixed by 
https://github.com/docker/docker/commit/34f4b197b8cec5b177797e343a5c89473ff1d2aa

vieux commented:
I confirm it fixes network:host :+1:

anusha-ragunathan commented on integration-cli/docker_cli_daemon_plugins_test.go:
```diff
@@ -105,7 +103,7 @@ func (s *DockerDaemonSuite) TestDaemonKillLiveRestoreWithPlugins(c *check.C) {
 	}
 
-	cmd := exec.Command("pgrep", "-f", "plugin-no-remove")
+	cmd := exec.Command("pgrep", "-f", "sample-volume-plugin")
```
Use pluginProcessName

anusha-ragunathan commented on plugin/store/store.go:
```diff
@@ -151,7 +151,12 @@ func (ps *Store) Get(name, capability string, mode int) (plugingetter.CompatPlug
+			}
+			// Plugin was found but it is disabled, so we should not fall back to legacy plugins
+			// but we should error out right away
+			return nil, ErrNotFound(fullName)
```
Its better to convey the correct error here. Something like `ErrNotEnabled`, so that user understand that enabling the plugin will resolve the error.

anusha-ragunathan commented on plugin/manager_linux.go:
```diff
@@ -27,10 +28,23 @@ func (pm *Manager) enable(p *v2.Plugin, force bool) error {
 	}
 
+	if p.NeedPropagation {
+		if err := syscall.Unmount(p.Rootfs, syscall.MNT_DETACH); err != nil {
```
why not do a lazy unmount, right after the mount in L36?

anusha-ragunathan commented on plugin/v2/plugin.go:
```diff
@@ -286,18 +302,51 @@ func (p *Plugin) GetTypes() []types.PluginInterfaceType {
+	if p.PluginObj.Config.Network.Type != "" {
+		// TODO: if net == bridge, use libnetwork controller to create a new plugin-specific bridge, bind mount /etc/hosts and /etc/resolv.conf look at the docker code (allocateNetwork, initialize)
+		if p.PluginObj.Config.Network.Type == "host" {
+			for i, n := range s.Linux.Namespaces {
```
Lets reuse `delNamespace` from oci_linux.go here.

anusha-ragunathan commented on plugin/v2/plugin.go:
```diff
@@ -322,6 +371,28 @@ func (p *Plugin) InitSpec(s specs.Spec, libRoot string) (*specs.Spec, error) {
+	for i, m := range s.Mounts {
+		if strings.HasPrefix(m.Destination, "/dev/") {
+			if _, ok := userMounts[m.Destination]; ok {
+				s.Mounts = append(s.Mounts[:i], s.Mounts[i+1:]...)
```
why cant plugin config have mount destinations in "/dev"?

anusha-ragunathan commented on api/types/plugin.go:
```diff
@@ -117,6 +113,23 @@ type PluginConfigInterface struct {
 
+// PluginConfigLinux plugin config linux
+// swagger:model PluginConfigLinux
+type PluginConfigLinux struct {
```
https://github.com/docker/docker/blob/master/docs/extend/config.md needs to be updated with these details.

tiborvass commented on plugin/v2/plugin.go:
```diff
@@ -309,7 +358,7 @@ func (p *Plugin) InitSpec(s specs.Spec, libRoot string) (*specs.Spec, error) {
 		}
 		if m.Source != "" && m.Type == "bind" {
-			fi, err := os.Lstat(filepath.Join(rootfs, m.Destination)) // TODO: followsymlinks
+			fi, err := os.Lstat(filepath.Join(p.Rootfs, string(os.PathSeparator), m.Destination)) // TODO: followsymlinks
```
you're right, this should be `string(os.PathSeparator) + m.Destination` to prevent `../`.

crosbymichael replied:
if you want to prevent `../` then you need to get the `Abs` of the path and compare.  lstat will follow the symlink but i'm not sure what you are trying to do here, its naive to look for a destination like this and i don't know why u have to look for the dest to create the source. 

the dest in the resulting container has to take all the previous mounts into account so i'm unsure that this will do what you want in the end

tiborvass replied:
@crosbymichael I think this code was trying to make sure the source directory is created.

anusha-ragunathan commented on pkg/system/devices_linux.go:
```diff
@@ -0,0 +1,87 @@
+	"path/filepath"
+	"strings"
+
+	"github.com/opencontainers/runc/libcontainer/configs"
```
Just looking at the imports, it seems like something that would belong in `runc/libcontainer`.

anusha-ragunathan commented on plugin/v2/plugin.go:
```diff
@@ -242,11 +251,18 @@ func (p *Plugin) ComputePrivileges() types.PluginPrivileges {
+	if m.Linux.DeviceCreation {
+		privileges = append(privileges, types.PluginPrivilege{
+			Name:        "device-creation",
+			Description: "",
```
`Description` for all privileges is left empty. Can we fill it in with descriptions?

anusha-ragunathan commented on plugin/v2/plugin.go:
```diff
@@ -322,6 +371,28 @@ func (p *Plugin) InitSpec(s specs.Spec, libRoot string) (*specs.Spec, error) {
+		rwm := "rwm"
+		s.Linux.Resources.Devices = []specs.DeviceCgroup{{Allow: true, Access: &rwm}}
+	}
+	for _, dev := range p.PluginObj.Config.Linux.Devices {
```
Can we reuse `setDevices` from oci_linux.go?

anusha-ragunathan replied:
Overall, plugin InitSpec should reuse daemon's container initspec code path

anusha-ragunathan commented on plugin/v2/plugin.go:
```diff
@@ -331,11 +402,22 @@ func (p *Plugin) InitSpec(s specs.Spec, libRoot string) (*specs.Spec, error) {
+
+	// TODO: what about duplicates?
+	// TODO: Should not need CAP_ prefix in manifest?
+	s.Process.Capabilities = append(s.Process.Capabilities, p.PluginObj.Config.Linux.Capabilities...)
```
Use `caps.TweakCapabilities()` with an empty cap drop list.

anusha-ragunathan commented:
This PR fixes https://github.com/docker/docker/issues/28131. Please include that in the commit, description.

anusha-ragunathan commented:
This PR introduces plugin support for device creation and host networking. It would be nice to get some unit test around it. I'm okay if this is done in a follow up PR.

Commit:
plugins: container-rootfs-relative paths

Legacy plugins expect host-relative paths (such as for Volume.Mount).
However, a containerized plugin cannot respond with a host-relative
path. Therefore, this commit modifies new volume plugins' paths in Mount
and List to prepend the container's rootfs path.

This introduces a new PropagatedMount field in the Plugin Config.
When it is set for volume plugins, RootfsPropagation is set to rshared
and the path specified by PropagatedMount is bind-mounted with rshared
prior to launching the container. This is so that the daemon code can
access the paths returned by the plugin from the host mount namespace.

Signed-off-by: Tibor Vass <tibor@docker.com>

vieux commented:
LGTM

anusha-ragunathan commented on plugin/v2/plugin.go:
```diff
@@ -308,12 +315,21 @@ func (p *Plugin) InitSpec(s specs.Spec, libRoot string) (*specs.Spec, error) {
+			Type:        "bind",
+			Options:     []string{"rbind", "rshared"},
+		},
+		types.PluginMount{
```
This block needs to be under `NeedPropagation` as well.

anusha-ragunathan commented on plugin/v2/plugin.go:
```diff
@@ -308,12 +315,21 @@ func (p *Plugin) InitSpec(s specs.Spec, libRoot string) (*specs.Spec, error) {
-		Type:        "bind",
-		Options:     []string{"rbind", "rshared"},
-	})
+	pathToMnt := filepath.Join(p.Rootfs, "mnt")
```
Can be shared with manager_linux.go L32.

cpuguy83 commented on pkg/plugins/plugins.go:
```diff
@@ -78,6 +78,12 @@ type Plugin struct {
 }
 
+// BasePath returns the path to which all paths returned by the plugin are relative to.
+// For legacy plugins, this always returns the host's root directory.
```
Just a nit, but s/legacy/v1, since everything becomes legacy at some point.

cpuguy83 commented on plugin/manager_linux.go:
```diff
@@ -27,10 +28,27 @@ func (pm *Manager) enable(p *v2.Plugin, force bool) error {
 	p.Restart = true
 	p.Unlock()
+
+	mnt := filepath.Join(p.Rootfs, "mnt")
```
If we're going to do this, then yes. My plugin just failed due to this as well. But I'm not sure we should really do this. Plugins can add the mount if they want to, no? Or is this just so they plugin author doesn't have to ask permission from the user since it's a guarantee?

cpuguy83 replied:
Seems like this is fixed now.

srust commented:
I'm confused what issue this is solving. In my plugin I am expecting to add a shared mount for `/some/path:/container/path` and mount volumes in the container at `/container/path`. The plugin Mount and List will respond with `/some/path/volume`. I am writing the manifest / config, so have full control over the paths that are being used by the plugin and volumes, so expect that the plugin should know what it's doing.

Is that not going to work for v2 plugins?

For what its worth, I've been doing the same thing in a containerized v1 plugin for over a year successfully.

cpuguy83 commented:
@srust I think the problem here is that the plugin author needs to know in advance how/where to setup mounts and also account for it in their own code.
It also means running 2 instances of the same plugin is (mostly) out of the question.

With this in a v2 plugin you always mount to the same location and that's namespaced by the plugin.

LK4D4 commented:
@tiborvass needs rebase

anusha-ragunathan commented:
Please take care of the following:
1. The static "/mnt" directory should be user configurable in config.json.
2. The unmount happens during plugin disable.
3. A test to be added for mount propagation. You can do something similar to https://github.com/opencontainers/runc/blob/509b1db98c98cf50612d5f08e21f4dd3d779fd41/libcontainer/integration/exec_test.go#L1289

cpuguy83 commented:
@anusha-ragunathan I also think this dir should be mounted outside of the rootfs so cases like a plugin upgrade won't overwrite data here.

anusha-ragunathan commented:
@cpuguy83 : good point. We could use parent `p.libRoot + pluginID` (personally prefer this) or `pRuntimeSourcePath` instead of `p.Rootfs` https://github.com/docker/docker/pull/26398/files#diff-ab0a5ea6d4b42eb51f9e05c2b536a89eR141

anusha-ragunathan commented:
Need a fresh pair of eyes reviewing this. @cpuguy83 ?

anusha-ragunathan commented on plugin/v2/plugin.go:
```diff
@@ -398,6 +402,7 @@ func (p *Plugin) InitSpec(s specs.Spec, libRoot string) (*specs.Spec, error) {
 	s.Process.Env = envs
 
 	s.Process.Capabilities = append(s.Process.Capabilities, p.PluginObj.Config.Linux.Capabilities...)
+	s.Process.Capabilities = append(s.Process.Capabilities, "CAP_SYS_ADMIN")
```
remove

anusha-ragunathan commented on plugin/manager.go:
```diff
@@ -124,6 +134,19 @@ func (pm *Manager) init() error {
+					if p.PluginObj.Config.PropagatedMount != "" {
+						// TODO: sanitize PropagatedMount and prevent breakout
+						p.PropagatedMount = filepath.Join(p.Rootfs, p.PluginObj.Config.PropagatedMount)
+						fmt.Println("yolo", p.PropagatedMount)
```
remove

anusha-ragunathan commented on plugin/manager.go:
```diff
@@ -124,6 +134,19 @@ func (pm *Manager) init() error {
+				if typ.Capability == "volumedriver" && typ.Prefix == "docker" && strings.HasPrefix(typ.Version, "1.") {
+					if p.PluginObj.Config.PropagatedMount != "" {
+						// TODO: sanitize PropagatedMount and prevent breakout
+						p.PropagatedMount = filepath.Join(p.Rootfs, p.PluginObj.Config.PropagatedMount)
```
Please see https://github.com/docker/docker/pull/26398#issuecomment-264586293

anusha-ragunathan commented on integration-cli/docker_cli_daemon_plugins_test.go:
```diff
@@ -197,10 +197,21 @@ func (s *DockerDaemonSuite) TestVolumePlugin(c *check.C) {
+		c.Log("plugin disabled")
+		mounts, err := mount.GetMounts()
+		c.Assert(err, checker.IsNil)
+		for _, mnt := range mounts {
```
We are simply printing out the mounts, which doesnt test much. Can we use `findmnt` to see if the mount propagated?

cpuguy83 commented:
Beyond the already commented items, code seems ok.

vdemeester commented on integration-cli/docker_cli_plugins_test.go:
```diff
@@ -32,7 +32,7 @@ func (s *DockerSuite) TestPluginBasicOps(c *check.C) {
 	c.Assert(err, checker.IsNil)
 
-	out, _, err = dockerCmdWithError("plugin", "remove", pNameWithTag)
+	out, _, _ = dockerCmdWithError("plugin", "remove", pNameWithTag)
```
Shouldn't this be `dockerCmd` instead ? Doesn't seem right to not care about error at all 😓

tiborvass replied:
@vdemeester I kept `err` and asserted it non-nil. Realized that `dockerCmd` will assume non-nil error...

anusha-ragunathan commented on plugin/manager.go:
```diff
@@ -124,6 +133,24 @@ func (pm *Manager) init() error {
+			for _, typ := range p.PluginObj.Config.Interface.Types {
+				if typ.Capability == "volumedriver" && typ.Prefix == "docker" && strings.HasPrefix(typ.Version, "1.") {
+					if p.PluginObj.Config.PropagatedMount != "" {
+						// TODO: sanitize PropagatedMount and prevent breakout
```
Is this still applicable? When does the mount leak?

cpuguy83 replied:
`../../../../etc`

anusha-ragunathan commented on api/types/plugin.go:
```diff
@@ -71,6 +71,10 @@ type PluginConfig struct {
 
+	// propagated mount
+	// Required: true
+	PropagatedMount string `json:"PropagatedMount"`
```
https://github.com/docker/docker/blob/master/docs/extend/config.md needs to be updated. Please do that alongside.

anusha-ragunathan commented:
LGTM (if tests pass)

cpuguy83 commented on plugin/manager.go:
```diff
@@ -141,6 +150,24 @@ func (pm *Manager) reload() error {
+				if typ.Capability == "volumedriver" && typ.Prefix == "docker" && strings.HasPrefix(typ.Version, "1.") {
+					if p.PluginObj.Config.PropagatedMount != "" {
+						// TODO: sanitize PropagatedMount and prevent breakout
+						p.PropagatedMount = filepath.Join(p.LibRoot, p.PluginObj.Config.PropagatedMount)
```
`filepath.Base(p.PluginObj.Config.PropagatedMount)`?

cpuguy83 replied:
Making sure I understand this... this is for mounting the propagated path to the plugin root (outside rootfs)?
Maybe even just use `mnt` instead of the mount path.

anusha-ragunathan replied:
@tiborvass : You should be using `filepath.Join(p.LibRoot, p.PluginObj.ID, p.PluginObj.Config.PropagatedMount)`. Else mounts from different plugins with the same name will collide. Another option is to fix `p.LibRoot` to include the pluginID (It should have been this way in the first place)

We dont want to use hard coded `mnt`. We want plugin devs to specify the mount that needs propagation.

cpuguy83 replied:
Right, but isn't this a singleton that's the "host" side of the mount?

anusha-ragunathan replied:
Yes. Having the same dir as the mount in the config could be less confusing. On the other hand, having a standard dir such as `mnt` is also reasonable. I'll let @tiborvass decide. Also, we dont have to do this in this PR.

tiborvass replied:
For start up reasons, this field is also set in `plugin/manager.go`. And I just realized it's set to something else. So I'll go with p.Rootfs for now.

Mrnamluxury reviewed (commented):
h
